jena-text/src/main/java/org/apache/jena/query/text/TextDatasetFactory.java

```diff
@@ -191,5 +213,50 @@ public static DatasetGraph createSolrIndex(DatasetGraph base, SolrServer server,
         TextIndex index = createSolrIndex(server, entMap) ;
         return create(base, index, true) ; 
     }
+ 
+    protected static class LuceneIndexMap {
+    	
+    	private static HashMap<Object, WeakReference<TextIndex>> map = new HashMap<Object, WeakReference<TextIndex>>();
+
+    	protected synchronized TextIndex get(Directory directory) {
+    		Object key = key(directory);
+    		WeakReference<TextIndex> ref = map.get(key);
+    		if (ref == null)
+    			return null;
+    		TextIndex result = ref.get();
+    		if (result == null)
+    			map.put(key, null);
+    		return result;
+    	}
+    	
+    	protected synchronized void put(Directory directory, TextIndex index) {
+    		map.put(key(directory), new WeakReference<TextIndex>(index));
+    	}
+    	
+    	protected synchronized void remove(Directory directory) {
+    		map.put(key(directory), null);
+    	}
+    	
+    	private Object key(Directory directory) {
+    		// uglyness alert
+    		if (directory instanceof FSDirectory) {
+    			return getFilePath( ( (FSDirectory) directory).getDirectory() );
+    		} else if (directory instanceof CompoundFileDirectory) {
+    			return getFilePath( ( (FSDirectory) directory).getDirectory() );
```

Comment:
Isn't this an invalid cast, you check for `instanceof CompoundFileDirectory` and then proceed to cast to `FSDirectory` which is going to fail

Unless `CompoundFileDirectory` is actally a type derived from `FSDirectory` in which case the previous branch of the `if` would have caught it already and this branch can never be reached